### PR TITLE
[Enchance] Add a probability parameter to Mosaic transform

### PR DESCRIPTION
**Pull request by RangeKing**
## Motivation

Add a probability parameter to the mosaic transform

## Modification

Add
```python
if random.uniform(0, 1) > self.prob:
            return results
```
in \_\_call__ of the mosaic transform.

## BC-breaking (Optional)

> Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

Not BC-breaking

## Use cases (Optional)

In config file,
```python
train_pipeline = [
    dict(type='Mosaic', img_scale=img_scale, pad_val=114, prob=0.5),
...
```

## Checklist

- [x] 1. Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] 2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] 3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
- [ ] 4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdet/datasets/pipelines/transforms.py [hunk 1]
```diff
@@ -2022,6 +2026,9 @@ def __call__(self, results):
             dict: Result dict with mosaic transformed.
         """
 
+        if random.uniform(0, 1) > self.prob:
```
Comment (hhaAndroid) [inline mmdet/datasets/pipelines/transforms.py:2029]: Please add unit test to prob.

**Comment by hhaAndroid** [comment]
@RangeKing Hello. do you have any progress?

File: mmdet/datasets/pipelines/transforms.py (hunk 1 above)
Comment (RangeKing) [reply mmdet/datasets/pipelines/transforms.py:2031]: > Please add unit test to prob.

I add an assertion test for invalid `prob` in `tests/test_data/test_pipelines/test_transform/test_transform.py`.
Does the unit test need to set different probabilities like `test_img_augment.py`?
https://github.com/open-mmlab/mmdetection/blob/e045607843f439f064791506257e9b4fdbf268c0/tests/test_data/test_pipelines/test_transform/test_img_augment.py#L138-L142

**Comment by hhaAndroid** [comment]
@RangeKing Please fix the conflicting files

**Comment by hhaAndroid** [comment]
@RangeKing Please fix the lint

File: mmdet/datasets/pipelines/transforms.py [hunk 2]
```diff
@@ -2002,15 +2004,20 @@ def __init__(self,
                  min_bbox_size=0,
                  bbox_clip_border=True,
                  skip_filter=True,
-                 pad_val=114):
+                 pad_val=114,
+                 prob=0.5):
```
Comment (hhaAndroid) [inline mmdet/datasets/pipelines/transforms.py:2008]:
```suggestion
                 prob=1.):
```

File: mmdet/datasets/pipelines/transforms.py (hunk 1 above)
Comment (hhaAndroid) [reply mmdet/datasets/pipelines/transforms.py:2032]: Yes

**Comment by RangeKing** [comment]
> @RangeKing Please fix the lint

@hhaAndroid I saw that the [lint check](https://github.com/open-mmlab/mmdetection/runs/5600909421?check_suite_focus=true#logs) was successful. But some build checks were not successful. I don't know much about the build error, for example, [build_cuda101 (3.7, 1.7.0+cu101)](https://github.com/open-mmlab/mmdetection/runs/5600910041?check_suite_focus=true#logs). I don't know how to fix it. Any suggestions, please.

File: mmdet/datasets/pipelines/transforms.py (hunk 2 above)
Comment (RangeKing) [reply mmdet/datasets/pipelines/transforms.py:2008]: image

ZwwWayne gave this suggestion that 0.5 might be better.

Comment (RangeKing) [reply mmdet/datasets/pipelines/transforms.py:2008]: I also think the default value set to 1 has the least impact.

Comment (hhaAndroid) [reply mmdet/datasets/pipelines/transforms.py:2008]: Change it to 1 because there cannot be bc breaking

**Comment by hhaAndroid** [comment]
> 

This is because you modified the logic to cause the unit test assert error.

File: tests/test_data/test_pipelines/test_transform/test_transform.py [hunk 3]
```diff
@@ -913,7 +918,7 @@ def test_mosaic():
     results['gt_labels'] = np.ones(gt_bboxes.shape[0], dtype=np.int64)
     results['gt_bboxes'] = gt_bboxes
     results['gt_bboxes_ignore'] = gt_bboxes_ignore
-    transform = dict(type='Mosaic', img_scale=(10, 12))
+    transform = dict(type='Mosaic', img_scale=(10, 12), prob=1.)
```
Comment (hhaAndroid) [inline tests/test_data/test_pipelines/test_transform/test_transform.py:921]: No changes required

**Review by hhaAndroid — APPROVED** [review]

**Review by ZwwWayne — APPROVED** [review]

**Comment by feixiang7701** [comment]
@RangeKing Hi. RandomAffine is often used with mosaic, if set the prob=0.5, the border of RandomAffine need to change together?
-----------------------------------------------------------------------------------------------------------------------------------------
border (tuple[int]): Distance from height and width sides of input image to adjust output shape. Only used in mosaic dataset.
            Default: (0, 0).
-----------------------------------------------------------------------------------------------------------------------------------------

**Comment by feixiang7701** [comment]
def __call__(self, results):
        img = results['img']
        height = img.shape[0] + self.border[0] * 2 
        width = img.shape[1] + self.border[1] * 2 #the height and width should be obtained from pipeline

**Comment by feixiang7701** [comment]
def __call__(self, results):
        img = results['img']
        height = img.shape[0] + self.border[0] * 2 
        width = img.shape[1] + self.border[1] * 2 #the height and width should be obtained from pipeline?

**Pull request merged**